tests/python/pants_test/util/test_objects.py

```diff
@@ -120,3 +168,139 @@ def test_unexpect_kwarg(self):
+    self.assertTrue(composed_decl.matches_value(3))
+    self.assertFalse(composed_decl.matches_value(type('asdf')))
+
+  def test_type_with_pred(self):
+    with self.assertRaises(TypeCheckError):
+      NonNegativeInt(value='asdf')
+
+    self.assertEqual(NonNegativeInt(value=3).value, 3)
+
+    with self.assertRaises(TypeCheckError):
+      NonNegativeInt(value=-1)
+
+
+class TypedDatatypeTest(BaseTest):
+
+  def test_class_construction(self):
+    # NB: typed_datatype subclasses declared at top level are the success cases
+    # here by not failing on import.
+
+    # If the type_name can't be converted into a suitable identifier, throw a
+    # ValueError.
+    with self.assertRaises(ValueError):
+      class NonStrType(typed_datatype(3, {'a': int})): pass
+
+    # This raises a TypeError because it doesn't provide a required argument.
+    with self.assertRaises(TypeError):
+      class NoFields(typed_datatype('NoFields')): pass
+
+    with self.assertRaises(TypedDatatypeClassConstructionError):
+      class NonDictFields(typed_datatype('NonDictFields', [
+          ('field', 'value'),
+      ])): pass
+
+    with self.assertRaises(TypedDatatypeClassConstructionError):
+      class NonTypeFields(typed_datatype('NonTypeFields', {'a': 3})): pass
+
+    with self.assertRaises(TypedDatatypeClassConstructionError):
+      class NonTypeUnionFields(typed_datatype('NonTypeUnionFields', {
+          'a': [str, 3],
+      })): pass
+
+    some_object = YetAnotherNamedTypedDatatype(
+      nothing_special=str('asdf'),
+      just_another_arg=3)
+    self.assertEqual(3, some_object.just_another_arg)
+
+    another_object = YetAnotherNamedTypedDatatype(
+      nothing_special=str('huh'),
+      just_another_arg=str('wow'))
+    self.assertEqual(str('wow'), another_object.just_another_arg)
+
+  def test_instance_construction(self):
+
+    some_val = SomeTypedDatatype(my_val=3)
+    self.assertIn('SomeTypedDatatype', repr(some_val))
+    self.assertIn('my_val', repr(some_val))
+    self.assertIn('3', repr(some_val))
+
+    union_val = UnionFieldTypedDatatype(an_arg=str('huh'))
+    self.assertIn('UnionFieldTypedDatatype', repr(union_val))
+    self.assertIn('an_arg', repr(union_val))
+    self.assertIn('huh', repr(union_val))
+
+    other_union_val = UnionFieldTypedDatatype(an_arg=3)
+    self.assertIn('UnionFieldTypedDatatype', repr(other_union_val))
+    self.assertIn('an_arg', repr(other_union_val))
+    self.assertIn('3', repr(other_union_val))
+
+    with self.assertRaises(TypeCheckError):
+      UnionFieldTypedDatatype(an_arg=[])
+
+    # no positional args are allowed
+    with self.assertRaises(TypedDatatypeInstanceConstructionError):
+      SomeTypedDatatype('hey')
+
+    # not providing all the fields
+    try:
+      SomeTypedDatatype()
+      self.fail("should have errored: not providing all constructor fields")
+    except TypedDatatypeInstanceConstructionError as e:
+      self.assertIn('my_val', str(e))
```

Comment:
I'd like to see more tests that check parts of the error messages. I think you could elide some of the comments like `# not providing all the fields` if there were an assertion that checked for that information.

One pattern that might be helpful is to use the context manager provided by `assertRaises`. For example
```python
with self.assertRaises(MyError) as cm:
  call()
self.assertIn('my error is the best error', cm.exception)
```

Then you can rewrite the above `try, fail, except, assert` as

```python
with self.assertRaises(TypedDatatypeInstanceConstructionError) as cm:
  SomeTypedDatatype()

self.assertIn('my_val', str(cm.exception))
self.assertIn('missing field', str(cm.exception)) # or something like it
```

Reply:
I really appreciate this feedback. I didn't make it clear in the OP, but I absolutely want to make sure error messages (and test cases) are deeply considered and rock solid for a feature like this. I didn't look deeply enough into assertRaises to realize that it yielded a context object and this seems really actionable. With respect to the code example you gave, I would probably do that and extract it into a helper method of the test class. The code samples you've provided are clear enough for me to act on.

Reply:
I canonicalized the `__str__()` and `__repr__()` of `TypedDatatype`, as well as the messages in the exceptions it raises, and added tests for all the failure cases in 133ba06. Let me know if this is what you were looking for.